src/app/icd/client/ICDRefreshKeyInfo.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ICDRefreshKeyInfo.h"
+#include "CheckInDelegate.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandPathParams.h>
+#include <app/InteractionModelEngine.h>
+#include <app/InteractionModelTimeout.h>
+#include <app/OperationalSessionSetup.h>
+
+namespace chip {
+namespace app {
+
+ICDRefreshKeyInfo::ICDRefreshKeyInfo(CheckInDelegate * apCheckInDelegate, const ICDClientInfo aICDClientInfo,
+                                     const ICDClientStorage * aICDClientStorage) :
+    mOnConnectedCallback(HandleDeviceConnected, this),
+    mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{
+    mpCheckInDelegate  = apCheckInDelegate;
+    mICDClientInfo     = aICDClientInfo;
+    mpICDClientStorage = const_cast<ICDClientStorage *>(aICDClientStorage);
```

Comment:
There can't be any const_cast here.  You don't control the caller; if that thing is in fact const, using it in a non-const manner will crash.  Please fix these APIs to make sense.

Reply:
Removed it and made both ICDClientInfo and ICDClientStorage non-const. ICDClientInfo will be used by the delegate in constructing ICDRefreshKeyInfo which needs to be modified on successful registration. ICDClientStorage has to be non-const to access the non-const function SetKey to update the new key to persistence.